codalab/worker_manager/worker_manager.py

```diff
@@ -0,0 +1,169 @@
+        self.args = args
+        self.codalab_manager = CodaLabManager()
+        self.codalab_client = self.codalab_manager.client(args.server)
+        self.staged_uuids = []
+        self.wait_for_progress = False  # Started a worker, waiting for something to happen
+        self.last_worker_start_time = 0
+        logger.info('Started worker manager.')
+
+    def get_worker_jobs(self):
+        """Return a list of `WorkerJob`s."""
+        raise NotImplementedError
+
+    def start_worker_job(self):
+        """Start a new `WorkerJob`."""
+        raise NotImplementedError
+
+    def run_loop(self):
+        while True:
+            self.run_one_iteration()
+            if self.args.once:
+                break
+            logger.debug('Sleeping {} seconds'.format(self.args.sleep_time))
+            time.sleep(self.args.sleep_time)
+
+    def run_one_iteration(self):
+        # Get staged bundles
+        keywords = ['state=' + State.STAGED] + self.args.search
+        if self.args.worker_tag:
+            keywords.append('request_queue=tag=' + self.args.worker_tag)
+        bundles = self.codalab_client.fetch(
+            'bundles', params={'worksheet': None, 'keywords': keywords, 'include': ['owner']}
+        )
+        new_staged_uuids = [bundle['uuid'] for bundle in bundles]
+        old_staged_uuids = self.staged_uuids
+        # Bundles that were staged but now aren't
+        removed_uuids = [uuid for uuid in old_staged_uuids if uuid not in new_staged_uuids]
+        if len(removed_uuids) > 0:
+            self.wait_for_progress = False
+        self.staged_uuids = new_staged_uuids
+        logger.info(
+            'Staged bundles [{}]: {}'.format(
+                ' '.join(keywords), ' '.join(new_staged_uuids) or '(none)'
+            )
+        )
+
+        # Get worker jobs
+        worker_jobs = self.get_worker_jobs()
+
+        # Print status
+        logger.info(
+            '{} staged bundles ({} removed since last time{}), {} worker jobs (min={}, max={})'.format(
+                len(new_staged_uuids),
+                len(removed_uuids),
+                ', waiting for >0 before launching worker' if self.wait_for_progress else '',
+                len(worker_jobs),
+                self.args.min_workers,
+                self.args.max_workers,
+            )
+        )
+
+        # There are two conditions under which we want more workers.
+        want_more_workers = False
+
+        # 1) We have fewer than min_workers.
+        if len(worker_jobs) < self.args.min_workers:
+            logger.info(
+                'Want to launch a worker because we are under the minimum ({} < {})'.format(
+                    len(worker_jobs), self.args.min_workers
+                )
+            )
+            want_more_workers = True
+
+        # 2) There is a staged bundle AND we have made progress running the
+        # previously staged bundles.
+        # To expand on this, if we just started a worker and we haven't made any
```

Comment:
There seems to be some room for error here. On the test deployment on SM, the gpu-worker-manager is stuck in a loop of no progress made while bundles are waiting staged. 

I think one plausible scenario for this is:
1. Worker manager is started, there are neither staged bundles nor worker jobs
2. Worker manager launches 1 worker job (to reach `min_workers`)
3. `wait_for_progress` is set to `True` since a new worker was launched
4. A run bundle is submitted and assigned to this newly created worker
5. Worker manager starts its new iteration, but isn't aware that its recently created worker picked up a bundle, because that bundle was never in its `staged_uuids` list
6. Another run bundle is submitted to Codalab and starts waiting in the `staged` state.
7. The worker manager refuses to launch another worker job despite the staged bundle because it doesn't think any progress was made on its first worker.
8. The second run bundle has to wait for the first run bundle to finish before it can be run.

The problem here is that a bundle might get submitted and immediately assigned to a new worker before the worker-manager can register it as a staged bundle first and then realize it's no longer staged (ie from the worker manager's perspective, its latest worker is doing nothing, and the staged bundles cannot be assigned to it for some mysterious reason).

Right now I'm trying to wrap my head around whether we can get to such a degenerate state some other way, but this is definitely one way the current system can get stuck.